Feat: Connect via peerID

Nkovaturient opened the pull request:
**Issue:**  #177 

- PR created as part and discussion, here -> [Issue](https://github.com/libp2p/universal-connectivity/issues/177#issuecomment-2630830629) 

**Aims:**

-  Example where two peers find each other with the peer-id only (in the best case with a dht for detecting the multiadders)

**Description:**
1)findByPeerId function aims to find through PeerStore or DHT Lookup
2)DHT response and peerResponse as fallback

**Fix needed:**

- peerInfo response is retrieving a list of mutiaddrs alongwith bootstrap addrs, however on clicking connect thru either of them, an error of  `NoValidAddressesError: The dial request has no valid addresses` or `"DHT lookup failed"`

- OUTPUT from PEERINFO obj(mainly thru Known Peer, not DHT)

```
addresses
: 
(10) [{…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}]
id
: 
Ed25519PeerId {type: 'Ed25519', multihash: Digest, publicKey: Ed25519PublicKey, string: undefined, Symbol(@libp2p/peer-id): true}
metadata
: 
Map(3) {'AgentVersion' => Uint8Array(64), 'last-dial-success' => Uint8Array(13), 'ProtocolVersion' => Uint8Array(0)}
peerRecordEnvelope
: 
Uint8Array(737) [10, 36, 8, 1, 18, 32, 87, 101, 178, 119, 212, 235, 230, 116, 127, 160, 63, 160, 129, 209, 158, 114, 12, 73, 56, 97, 94, 174, 237, 156, 100, 176, 89, 177, 213, 203, 191, 239, 18, 2, 3, 1, 26, 242, 4, 10, 38, 0, 36, 8, 1, 18, 32, 87, 101, 178, 119, 212, 235, 230, 116, 127, 160, 63, 160, 129, 209, 158, 114, 12, 73, 56, 97, 94, 174, 237, 156, 100, 176, 89, 177, 213, 203, 191, 239, 16, 199, 222, 166, 171, 183, 153, 254, 142, 24, 26, 50, 10, 48, 4, …]
protocols
: 
(10) ['/floodsub/1.0.0', '/ipfs/id/1.0.0', '/ipfs/id/push/1.0.0', '/ipfs/kad/1.0.0', '/ipfs/ping/1.0.0', '/libp2p/circuit/relay/0.2.0/hop', '/libp2p/circuit/relay/0.2.0/stop', '/meshsub/1.0.0', '/meshsub/1.1.0', '/meshsub/1.2.0']
tags
: 
Map(7) {'kad-close' => {…}, 'kad-peer' => {…}, 'keep-alive-circuit-relay' => {…}, 'keep-alive-kad-dht' => {…}, 'universal-connectivity' => {…}, …}
```

![Screenshot (476)](https://github.com/user-attachments/assets/65033722-75bf-4702-9a46-ec380dc4f45f)


![Screenshot (476)](https://github.com/user-attachments/assets/e510347d-de2b-4d13-b448-286eded62498)

- why DHT lookup is failing?
- How to fetch peerInfo by entering peerId of  known connections with valid addrs

2color commented on js-peer/src/lib/libp2p.ts:
```diff
@@ -85,119 +118,317 @@ export async function startLibp2p(): Promise<Libp2pType> {
+      }
+
+      // Step 2: Find the peer using DHT lookup with retries
+      let event: QueryEvent | undefined;
```
Instead of using the DHT service, you should use `libp2p.peerRouting.findPeer(peerId)` which abstracts from the specific peerRouting implementations. This way, it can benefit from both delegated routing and the DHT.

I would also reduce the scope of this PR and not introduce the kadDHT.

2color replied:
See https://github.com/ipfs-shipyard/www-helia-identify/blob/main/src/index.ts#L59-L86 for how to simplify.

You probably don't even need the onPorgress hook from `findPeer`.

Nkovaturient replied:
alryt will implement and fix as desired

2color reviewed (commented):
Thanks for the PR, Neha.

Before a proper review:
- Can you please run prettier on the files you edited in this PR to format changes (I've opened https://github.com/libp2p/universal-connectivity/pull/211 to address this). A lot of changes in this PR are code style changes adding unnecessary noise that makes reviewing harder.
- Can you please remove any commented out code that isn't relevant for this.

Nkovaturient commented:
> Thanks for the PR, Neha.
> 
> Before a proper review:
> 
> * Can you please run prettier on the files you edited in this PR to format changes (I've opened [js-peer: add prettier for consistent styling #211](https://github.com/libp2p/universal-connectivity/pull/211) to address this). A lot of changes in this PR are code style changes adding unnecessary noise that makes reviewing harder.
> * Can you please remove any commented out code that isn't relevant for this.

Roger that! @2color

Nkovaturient commented:
Hey, @2color 
wanna update you with recent commits:-

- ran prettier on files and code changes
- modified kadDHT to peerRouting methods-indeed an effective way of finding peers
- updated the peerFind handler in `index.tsx` 
![Screenshot (486)](https://github.com/user-attachments/assets/d94594df-73b3-4f6a-bfef-7d8f6a34aa4c)

Nkovaturient commented:
Hey, kindly review and provide guidance--
 **summary of the key changes made for peer connectivity:**

> Basically `findBypeerID` is fetching  a list of multiaddrs --> in `peer-maddr.tsx` -we are filtering it with valid transport stack --> then trying to connect with either of them` handleConnect`

• **Connection Handling**:
  - Fixed connection object creation to use actual libp2p connection
  - Updated connection type interface to match simplified structure
  - Added proper connection state management

• **Transport Protocol Validation**:
  - Added checks for WebTransport + certhash
  - Added checks for WebRTC + QUIC + certhash
  - Added external address validation

• **Connection Flow**:
  - Parse multiaddr to extract peerID
  - Validate transport stack
  - Attempt connection via connectToMultiaddr
  - Store successful connection in state
  - Clear resolved multiaddrs after connection
  - Handle connection errors with user feedback

• **UI Improvements**:
  - Added connection priority system
  - Show transport protocol badges
  - Display connection status indicators
  - Added loading states during connection

![Screenshot (487)](https://github.com/user-attachments/assets/f49efb60-3572-4d32-ae3a-eef394878b6d)
![Screenshot (488)](https://github.com/user-attachments/assets/839f407d-48b1-42ac-b719-172d0118a224)

![Screenshot (489)](https://github.com/user-attachments/assets/73b0e200-e826-4c5e-968d-c9c2e1a776fd)

Nkovaturient commented:
@2color any updates/guidance?

2color commented on js-peer/src/lib/libp2p.ts:
```diff
@@ -68,7 +68,14 @@ export async function startLibp2p(): Promise<Libp2pType> {
-        list: bootstrapAddrs,
+        list: [
+          `${bootstrapAddrs}`,
+          '/ip6/2604:1380:4642:6600::3/udp/9095/quic-v1/webtransport/certhash/uEiAFmismVS4uGGz9zF8yLRC10wtqPciwcBD1BuAch4sX3A/certhash/uEiBEvL3ao0UqfMSkj2JCOvjG_4BEiiEnjFr7qmDPALgG5Q',
```
The bootstrapper maddrs shouldn't be hardcoded. These are only valid for ~4 weeks, as the certificates rotate automatically. This is why we go through the trouble of first resolving the PeerID to its latest multiaddrs

Nkovaturient replied:
I am **only** able to connect thru it by  bypassing the ` transport-protocol` and `Couldnt find any bootstrapper addrs` err with this explicit definition--however i shall remove it on my latest commit and changes

2color commented on js-peer/src/lib/libp2p.ts:
```diff
@@ -102,7 +109,10 @@ export async function startLibp2p(): Promise<Libp2pType> {
   libp2p.addEventListener('peer:discovery', (event) => {
     const { multiaddrs, id } = event.detail
-
+    let accioPeers = []
```
`accioPeers` can be removed

2color commented on js-peer/src/pages/index.tsx:
```diff
@@ -16,6 +17,48 @@ export default function Home() {
+
+    setLoading(true)
+    try {
+      const peerInfo = await findPeerById(libp2p)(peerIdInput.trim())
```
This can be radically simplified by first parsing the string into a PeerID (see https://github.com/ipshipyard/ipns-inspector/blob/c34d9b54c30be34eab10cb8ae8e5eca3721485d6/lib/peer-id.ts#L13-L22 for inspiration) and then just calling `libp2p.dial(peerID)`

The `dial` of js-libp2p will handle peer routing by using the the connection manager which calls the dial queue which looks up the peer store and if not found will use peer routing to resolve the PeerID:
- https://github.com/libp2p/js-libp2p/blob/b030eadb0c75db30d199cf2b261c14c23ce0a453/packages/libp2p/src/connection-manager/index.ts/#L489-L561
- https://github.com/libp2p/js-libp2p/blob/b030eadb0c75db30d199cf2b261c14c23ce0a453/packages/libp2p/src/connection-manager/dial-queue.ts#L138
- https://github.com/libp2p/js-libp2p/blob/b030eadb0c75db30d199cf2b261c14c23ce0a453/packages/libp2p/src/connection-manager/dial-queue.ts/#L315

2color commented on js-peer/src/pages/index.tsx:
```diff
@@ -137,6 +180,58 @@ export default function Home() {
                   </>
                 ) : null}
               </div>
+              {/* CONNECT BY PEER-ID */}
```
This could probably move into a separate component.

2color commented on js-peer/src/lib/libp2p.ts:
```diff
@@ -114,6 +124,62 @@ export async function startLibp2p(): Promise<Libp2pType> {
 }
 
+/* ability to Connect via peer-ID */
+export const findPeerById = (libp2p: Libp2pType) => async (peerIdStr: string) => {
```
See https://github.com/libp2p/universal-connectivity/pull/210/files#r1952594020

This function is redundant. This logic is already implemented in js-libp2p.

2color commented on js-peer/src/components/peer-list.tsx:
```diff
@@ -1,25 +1,43 @@
 
 interface PeerListProps {
-  connections: Connection[]
+  connections: Array<{
```
What's the reason for not using the `Connection` type here?

Nkovaturient replied:
since I was hovering between coding ui in `peer-list.ts` and `handlePeerId` handler in` index.js,` hence I defined the data inline shape temporarily as to understand and extract the properties... however, yes, its not required, I will make amends.

2color commented on js-peer/src/components/peer-maddr.tsx:
```diff
@@ -0,0 +1,169 @@
+  setError: (error: string | null) => void
+}
+
+function MaddrItem({ addr, setResolvedMultiaddrs, setError }: MaddrItemProps) {
```
A lot of this is redundant and is in js-libp2p.

2color commented on js-peer/src/context/ctx.tsx:
```diff
@@ -28,6 +34,7 @@ let loaded = false
 export function AppWrapper({ children }: WrapperProps) {
   const [libp2p, setLibp2p] = useState<Libp2pType | undefined>(undefined)
   const [error, setError] = useState('')
+  const [connections, setConnections] = useState<Connection[]>([])
```
If you're going to move connections to the context state here, also move the useEffect hook from the [pages/index.tsx](https://github.com/libp2p/universal-connectivity/pull/210/files#diff-da511c52b901e9c1f5a7c359c4c36eefa3376fe6ac7538242390c660bc47a684R63-R75) and replace other users of connections with the context state.

2color reviewed (changes requested):
This is starting to take shape nicely. See comments.

Main points:
- Please remove redundant logic that's already in js-libp2p as explained in https://github.com/libp2p/universal-connectivity/pull/210/files#r1952594020 
- Try to keep the new components as UI focused (with as little libp2p specific logic as possible) 
- See comment about the libp2p context.

Nkovaturient commented:
@2color Thanks a bunch for reviewing it.
 I shall get back to you with an update.

Nkovaturient commented:
Hello there, @2color 
I have updated with the following changes:-

- Remove the redundant findPeerById function and use libp2p.dial() directly
- Move connection state management to the context
- Create a dedicated PeerConnect component for the UI
- Simplify the connection logic by leveraging js-libp2p's built-in peer routing
- Remove the unnecessary peer multiaddr list component since we're connecting directly via PeerID

https://www.loom.com/share/8d3333d1d316419b8a882e4cb4f648f3?sid=b65f2e97-a290-424a-a0ad-1a6b608965c0

2color commented:
Thanks @Nkovaturient. 

I just took a look at this and think we can simplify this further. Instead of adding a new component to connect to PeerIDs, let's repurpose the existing `multiaddr` input and also allow PeerIDs there too. 

In the handler (what's now called `handleConnectToMultiaddr`), you can use heuristics (like inspecting the first char `/`) or just try parsing as PeerID and if that fails, try creating a Multiaddr.

Final note, before committing, please run `npm run format`.

Nkovaturient commented:
Sure @2color , that would be an efficient approach, lemme modify it as desired

Nkovaturient commented:
Hey @2color , why was the `bootstrap` module removed from main? 
Albeit short-term, It was an efficient module for anyone getting started with uc and avoiding unnecessary err of `no addresses discovered`

Nkovaturient commented:
https://github.com/libp2p/universal-connectivity/pull/225

Got it, okay 👍

2color commented on js-peer/src/components/peer-list.tsx:
```diff
@@ -40,26 +39,21 @@ function Peer({ connection }: PeerProps) {
         <div className="min-w-0 flex-auto">
-          <p className="text-sm font-semibold leading-6 text-gray-900">
-            {connection.remotePeer.toString()}{' '}
-            {connection.remoteAddr.protoNames().includes('webrtc') ? <Badge color="indigo">P2P Browser</Badge> : null}
```
Any reason you've removed the browser badge. This is useful for denoting direct browser connections.

Nkovaturient replied:
to keep the original ui unchanged.. however, I can revert it as such.

2color replied:
Not sure what you mean by original UI, since this was in the original UI.

Nkovaturient replied:
ah! my bad I misinterpreted it with another component.

Nkovaturient commented on js-peer/src/context/ctx.tsx:
```diff
@@ -14,9 +14,15 @@ export type Libp2pType = Libp2p<{
+export const libp2pContext = createContext<{
+  libp2p: Libp2pType
+  connections: Connection[]
+  setConnections: (connections: Connection[]) => void
```
yeah fs

Nkovaturient commented:
@2color ,  websockets connections are failing to establish connections on Chrome and Edge browser. I have tried refreshing a lot but to no avail ...??

![Screenshot (516)](https://github.com/user-attachments/assets/c8e20160-fd9c-4bb4-845a-2ffc7d13df4b)

![Screenshot (517)](https://github.com/user-attachments/assets/b177ebd7-3be5-4fc2-8f26-8600c40abe8a)

2color commented:
@Nkovaturient Are you getting these errors with the main branch too?

Nkovaturient commented:
> @Nkovaturient Are you getting these errors with the latest main branch too?
> 
> Can you also share the result of https://test-ipv6.com/? I pushed a fix ([34b6b35](https://github.com/libp2p/universal-connectivity/commit/34b6b356a702755f2213c92dc6f381ee023b3512)) that may cause be related to this.

Yeah, I have updated with the main branch yet the same errors. pkg update and fix ([34b6b35](https://github.com/libp2p/universal-connectivity/commit/34b6b356a702755f2213c92dc6f381ee023b3512)) seems okay, I have not been able to open the server since the bootstraplistenaddr were  removed and replaced by ipv4 and websockets nonetheless the relaylistenaddr

ipv6 tests
![Screenshot (518)](https://github.com/user-attachments/assets/87330669-ac4c-44f1-a3ac-618eeb853b97)

2color commented:
The latest changes undo the functionality intended to be introduced by this PR

Nkovaturient commented:
> The latest changes undo the functionality intended to be introduced by this PR

Kindly check is it up to the mark?

Nkovaturient commented:
@2color how to address this build err?
```
Deploy IPS Error: Either Storacha credentials ( and ) or IPFS Cluster credentials (, , and ) or Kubo credentials ( and ) must be configured
Error: Process completed with exit code 1.
```
